### ISSUE TITLE
Reactive Functional Programming

### ISSUE THREAD

**Issue by MattiaPrimavera** [slapperwan/gh4a#692]
Hi guys, 

did you already consider moving to introduce the code base to the **reactive** world, using the some of the following **Rx** libraries ? 

- [RxJava2](https://github.com/ReactiveX/RxJava)
- [RxPermissions](https://github.com/tbruyelle/RxPermissions) 
- [RxActivityResult](https://github.com/VictorAlbertos/RxActivityResult) 
- [RxAndroid](https://github.com/ReactiveX/RxAndroid)
- [RxBindings](https://github.com/JakeWharton/RxBinding)

**Comment by maniac103** [comment]
Yes, (at least) I did consider it. I don't think it provides enough additional value to justify rewriting the whole app + the bindings, though. I'm not even sure I'd use if I were to rewrite the app from scratch - the loader infrastructure simply seems sufficient for what we're doing.

In terms of functional programming, I'd consider making the app use a library for the Java8 stream API (e.g. https://github.com/aNNiMON/Lightweight-Stream-API) more important/useful.

@Tunous Any thoughts from your side?

**Comment by Tunous** [comment]
Even though the current codebase is dated I find it very easy to work with. I don't see any huge issues in how the app performs that would justify the rewrite. Rx is very nice but there is no reason to use it.

**Comment by MattiaPrimavera** [comment]
It's not issue-related but more an 'enhancement' purpose. Code would be clearified since all async operations would pass through the unique interface **Observer**, this would avoid wiring lots of **AsyncTasks** and **Loaders**, which split logic into multiple callbacks. 

**rxPermissions** and **reactivity result** eliminate the need to override **permissions** and **result** callbacks, does the developer does not always have to move among callbacks to follow the application flow,

**Comment by maniac103** [comment]
> It's not issue-related but more an 'enhancement' purpose. Code would be clearified since all async operations would pass through the unique interface Observer, this would avoid wiring lots of AsyncTasks and Loaders, which split logic into multiple callbacks.

There is no logic split here. The Loader subclasses are the equivalent of `Observable`, while the loader callbacks are the equivalent of `Observer`. The AsyncTasks for saving stuff (which aren't related to the loaders) are a combination of both `Observer` and `Observable`, so if anything, we'd need to split logic there.

RxPermissions, RxBindings etc. just wrap the Android API into Rx-like API, but as I mentioned before: Looking at potential contributors, I prefer staying close to the paradigms used in the Android API. (Will a potential contributor used to Rx, but not Android, be able to contribute? Likely not. Will a potential contributor used to Android, but not Rx, be able to contribute? Yes, if we follow standard Android design patterns.)

**Comment by MattiaPrimavera** [comment]
> I prefer staying close to the paradigms used in the Android API. (Will a potential contributor used to Rx, but not Android, be able to contribute? Likely not. Will a potential contributor used to Android, but not Rx, be able to contribute? Yes, if we follow standard Android design patterns.) 

I agree on this one, even if using some of the modern libraries and give some freshness to the code base does not forcefully mean **less contributors**, since I think a lot of today's android developers are used to tools like **retrofit2** and **rxJava**. 

With the premise that I find this is a great project :) I'm building architectures with **retrofit2** (with **okHttp** client) and **rxJava2** and I find that pretty concise and clear : 

**Retrofit** types the API and automatically convert the json response (through **gson** library) and cast it into a **Model** class. **RxJava** allow simply declaring that data has to be handled on a background thread, without the need of implementing all **Loaders** boilerplate code ... I personally find this pretty cool ... 

```java
Observable<Data> data = networkManager.getService()
  .compose(RxTools.applySchedulers())
  .subscribe(result -> {
     // do something on result 
  }
```

Anyway this project counts **56 loaders** ... among which a lot of code is pretty much similar ... shouldn't this be abstracted ? Through java **genericity** if you guys want to keep on without **rxJava2** ;) 

```
    @Override
    public List<RepositoryContents> doLoadInBackground() throws IOException {
        ContentsService contentService = (ContentsService)
                Gh4Application.get().getService(Gh4Application.CONTENTS_SERVICE);
        return contentService.getContents(new RepositoryId(mRepoOwner, mRepoName), mPath, mRef);
    }

    @Override
    public List<Contributor> doLoadInBackground() throws IOException {
        RepositoryService repoService = (RepositoryService)
                Gh4Application.get().getService(Gh4Application.REPO_SERVICE);
        return repoService.getContributors(new RepositoryId(mRepoOwner, mRepoName), true);
    }

    @Override
    public List<Download> doLoadInBackground() throws IOException {
        DownloadService downloadService = (DownloadService)
                Gh4Application.get().getService(Gh4Application.DOWNLOAD_SERVICE);
        return downloadService.getDownloads(new RepositoryId(mRepoOwner, mRepoName));
    }

    @Override
    public List<Gist> doLoadInBackground() throws IOException {
        GistService gistService = (GistService)
                Gh4Application.get().getService(Gh4Application.GIST_SERVICE);
        return gistService.getGists(mUserName);
    }
```

The number of **AsyncTask** classes I count is **36** instead ... I'm not saying **rxJava2** would work better or give better results, but I'm convinced it would help reducing boilerplate code ... I'm open to discussion and curious on your thoughts of course, if my considerations do not bother :)

**Comment by MattiaPrimavera** [comment]
What I meant for **logic split** among multiple **callbacks** can be easily seen with android way of handling **runtime permissions** or **activity results**, for example : 

```java
    @Override
    public void onItemClick(Issue issue) {
        String[] urlPart = issue.getUrl().split("/");
        Intent intent = issue.getPullRequest() != null
                ? PullRequestActivity.makeIntent(getActivity(), urlPart[4], urlPart[5], issue.getNumber())
                : IssueActivity.makeIntent(getActivity(), urlPart[4], urlPart[5], issue.getNumber());
        startActivityForResult(intent, REQUEST_ISSUE);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_ISSUE) {
            if (resultCode == Activity.RESULT_OK) {
                super.onRefresh();
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }
```

In the example above we start an activity for getting a result back in `onItemClick` method, while result is handled in `onActivityResult` one ... `RxActivityResult` would give something like ... 

```java
RxActivityResult.on(this).startIntent(INTENT_TO_LAUNCH)
        .subscribe(result -> {
            Intent data = result.data();
            int resultCode = result.resultCode();

            // the requestCode using which the activity is started can be received here.
            int requestCode = result.requestCode();

            if(requestCode == YourActivity.YOUR_REQUEST_CODE) {
                // Do Something
            }

            if (resultCode == RESULT_OK) {
                result.targetUI().showImage(data);
            } else {
                result.targetUI().printUserCanceled();
            }
        });
```

Starting the **activity** and getting its result would this way be contained in the same method, avoiding the developer to have to follow different callbacks to act on **permissions** or **activity** results.

**Comment by MattiaPrimavera** [comment]
Another disadvantage coming up to my mind is that all these `AsyncTasks` are often treated as `internal` **private classes** in order to keep the possibility of calling the **parent** class **methods**, without having to **pass** it the result through a callback, if we were sorting the **AsyncTask** out of the hosting fragment. 

Isn't that what happens for `ReviewFragment` containing `ReplyTask` and `DeleteCommentTask` (both making a call at the **fragment** `reloadComments` method) ? 

With **Observables**, an **Async operation** can be declared in a **separate file** avoiding **internal classes**, its declaration will be lighter (using `lambda` syntax) than writing **36** different classes, either **internal** or implementing **custom callbacks** to pass the data to the calling fragment.

**Comment by Tunous** [comment]
I think you made your point. I personally have nothing against rxjava itself. I've actually happily used it in other projects and I can see how it could benefit further development of the app.

Although, I don't think the benefit would be that huge since it's easy to work with the current codebase. That, and the fact that it would require a lot of work to update the app are the reasons I'm against this.

I'm not signing up for the rewrite myself. 

But, if what you want to do is to contribute these changes **yourself** then I'm ok with that. As long as it won't remove existing functionality or heavily disturb in development of new functionality.

**Comment by MattiaPrimavera** [comment]
I'd be happy to contribute on this :) 

Could you give me a little feedback on the idea ? I made a **pull-request** restrained to **RepositoryFragment**, still **work-in-progress** ...  but it runs and the readme is correctly downloaded + images within filled ;)

https://github.com/slapperwan/gh4a/pull/698

**Comment by Tunous** [comment]
Implemented in #778
